lib/telemetry_registry.ex

```diff
@@ -0,0 +1,253 @@
+  for users to enhance spans with attributes created from telemetry event measurements and metadata.
+  """
+
+  @typedoc """
+  An application to discover events from.
+  """
+  @type application() :: :telemetry_registry.application()
+
+  @typedoc """
+  A tuple containing the telemetry event, the module in which it was declared, and event definition meta.
+  `{:telemetry.event_name(), module(), t:event_meta()`
+  """
+  @type event() :: :telemetry_registry.event()
+
+  @typedoc """
+  An event definition is composed of an event, description, measurements description, and metadata description.
+  """
+  @type event_definition() :: :telemetry_registry.event_definition()
+
+  @typedoc """
+  A description of what the event represents and when it is emitted.
+  """
+  @type event_description() :: :telemetry_registry.event_description()
+
+  @typedoc """
+  A string representation of the measurements emitted with the event. This should resemble a typespec but is
+  not limited to the typespec format, i.e. you can include a comment on which unit a value is in. The objective
+  is to inform users.
+  """
+  @type event_measurements() :: :telemetry_registry.event_measurements()
+
+  @typedoc """
+  A string representation of the metadata emitted with the event. This should resemble a typespec but is
+  not limited to the typespec format, i.e. you can include comments or example values. The objective
+  is to inform users what is available.
+  """
+  @type event_metadata() :: :telemetry_registry.event_metadata()
+
+  @typedoc """
+  A map of event definition meta for an event containing the event, measurements, and metadata descriptions
+  if the event was declared with an event definition. Otherwise, this value will be an empty map.
+  """
+  @type event_meta() :: :telemetry_registry.event_meta()
+
+  @typedoc """
+  A list of spannable events known to the registry in the format of `{event_prefix, event_suffixes}`. For
+  example, given events `[:my_app, :request, :start], [:my_app, :request, :stop], [:my_app, :request, :exception]`
+  a span can be created from the `:start` -> `:stop` or the `:start` -> `:exception` events. These are aggregated
+  as a spannable event `{[:my_app, :request], [:start, :stop, :exception]}`.
+  """
+  @type spannable_event() :: :telemetry_registry.spannable_event()
+
+
+  defmacro __using__(_opts) do
+    quote do
+      import unquote(__MODULE__), only: [format_telemetry_docs: 1, telemetry_event: 1]
+
+      Module.register_attribute(__MODULE__, :telemetry_event,
+        accumulate: true,
+        persist: true
+      )
+
+      Module.register_attribute(__MODULE__, :telemetry_docs,
+        accumulate: true,
+        persist: true
+      )
+    end
+  end
+
+  @doc """
+  Declares a telemetry event. Accepts a telemetry event definition or telemetry event name. A full
+  definition is required for complete documentation and is highly encouraged.
+  """
+  defmacro telemetry_event(event) do
```

Comment:
I'm torn to be honest. If I don't preload the docs, it isn't necessary. However, it also enables adding validation for Elixir users (noted that I haven't added that yet). This could be extended to Erlang users via an Erlang macro, though I'm not sure they'd use it.

It's not necessary that they use this macro. The only "requirement" is adding the `use TelemetryRegistry` to set things up for using the attributes.

Reply:
Leaving this as a macro for future use (validation and such). Users are free to declare attributes by hand.